src/Induction/NoInfiniteDescent.agda

```diff
@@ -0,0 +1,221 @@
+
+    wf⇒infiniteDescent : WellFounded _<_ → InfiniteDescent
+    wf⇒infiniteDescent wf = acc⇒infiniteDescent (wf _)
+
+    acc⇒noInfiniteDescent : ∀[ Acc _<_ ⇒ ¬_ ∘ P ]
+    acc⇒noInfiniteDescent acc[x] px = Lemmas∩.acc⇒noInfiniteDescent acc[x] (px , acc[x])
+
+    wf⇒noInfiniteDescent : WellFounded _<_ → ∀[ ¬_ ∘ P ]
+    wf⇒noInfiniteDescent wf = acc⇒noInfiniteDescent (wf _)
+
+
+------------------------------------------------------------------------
+-- Further corollaries: assume _<⁺_ descent only for Acc _<⁺_  elements
+
+module FurtherCorollaries {_<_ : Rel A r} (P : Pred A ℓ) where
+
+  private
+
+    _<⁺_ = TransClosure _<_
+
+    module ID⁺ = InfiniteDescent _<⁺_ P
+
+    DescentAt⁺   = ID⁺.DescentAt
+    Descent⁺     = ID⁺.Descent
+
+  Acc⇒Descent⁺ : Set _
+  Acc⇒Descent⁺ = ∀[ Acc _<_ ⇒ DescentAt⁺ ]
+
+  InfiniteDescendingSequence⁺ : (f : ℕ → A) → Set _
+  InfiniteDescendingSequence⁺ f = ∀ {m n} → m ℕ.< n → f n <⁺ f m
+
+  InfiniteSequence⁺_From_ : (f : ℕ → A) → Pred A _
+  InfiniteSequence⁺ f From x = f zero ≡ x × InfiniteDescendingSequence⁺ f
+
+  InfiniteDescentAt⁺ : Pred A _
+  InfiniteDescentAt⁺ x = P x → ∃[ f ] Π[ P ∘ f ] × InfiniteSequence⁺ f From x
+
+  Acc⇒InfiniteDescent⁺ : Set _
+  Acc⇒InfiniteDescent⁺ = ∀[ Acc _<_ ⇒ InfiniteDescentAt⁺ ]
+
+  InfiniteDescent⁺ : Set _
+  InfiniteDescent⁺ = ∀[ InfiniteDescentAt⁺ ]
+
+  module _ (descent⁺ : Acc⇒Descent⁺) where
+
+    private
+      descent : ID⁺.Acc⇒Descent
+      descent = descent⁺  ∘ (accessible⁻ _)
+
+      module Corollaries⁺ = Corollaries _ descent
+
+      sequence⁺ : ∀ {f} →
+                  ID⁺.InfiniteDescendingSequence f → InfiniteDescendingSequence⁺ f
+      sequence⁺ {f} seq[f] = seq⁺[f]′ ∘ ℕ.<⇒<′
+        where
+        seq⁺[f]′ : ∀ {m n} → m ℕ.<′ n → f n <⁺ f m
+        seq⁺[f]′ ℕ.<′-base        = seq[f] _
+        seq⁺[f]′ (ℕ.<′-step m<′n) = seq[f] _ ++ seq⁺[f]′ m<′n
+
+
+    acc⇒infiniteDescent⁺ : Acc⇒InfiniteDescent⁺
+    acc⇒infiniteDescent⁺ acc[x] px
+      with f , Π[P∘f] , f0≡x , sequence[f] ← Corollaries⁺.acc⇒infiniteDescent (accessible _ acc[x]) px
+         = f , Π[P∘f] , f0≡x , sequence⁺ sequence[f]
+
+    wf⇒infiniteDescent⁺ : WellFounded _<_ → InfiniteDescent⁺
+    wf⇒infiniteDescent⁺ wf = acc⇒infiniteDescent⁺ (wf _)
+
+    acc⇒noInfiniteDescent⁺ : ∀[ Acc _<_ ⇒ ¬_ ∘ P ]
+    acc⇒noInfiniteDescent⁺ = Corollaries⁺.acc⇒noInfiniteDescent ∘ (accessible _)
+
+    wf⇒noInfiniteDescent⁺ : WellFounded _<_ → ∀[ ¬_ ∘ P ]
+    wf⇒noInfiniteDescent⁺ = Corollaries⁺.wf⇒noInfiniteDescent ∘ (wellFounded _)
+
+
+------------------------------------------------------------------------
+-- Finally:  the (classical) 'no smallest counterexample' principle itself
+
+module _ {_<_ : Rel A r} {P : Pred A ℓ} (stable : Stable P) where
+
+  open FurtherCorollaries {_<_ = _<_} (∁ P)
+    using (acc⇒noInfiniteDescent⁺; wf⇒noInfiniteDescent⁺)
+    renaming (Acc⇒Descent⁺ to NoSmallestCounterExample)
+
+  acc⇒noSmallestCounterExample : NoSmallestCounterExample → ∀[ Acc _<_ ⇒ P ]
```

Comment:
It's weird to have lemmas exported from this module that refer to the name `NoSmallestCounterExample` and for that name to not be exported from this module?

Reply:
Ha! Good catch!

Reply:
Is the name exported correctly now?